---
title: "GenomeProt Report"
date: "`r format(Sys.Date())`"
output: html_document
layout: single
params:
  directory:
    value: x
  file:
    value: x
---


```{r setup, include=FALSE, warning=FALSE}
library(ggplot2)
library(DESeq2)
library(gplots)
library(dplyr)
library(RColorBrewer)
library(data.table)
options(scipen=999)

df <- read.csv(paste0(params$directory, "/", params$file))

#df <- fread("/Users/josieg/Documents/peptide_info.csv")

bad_pep <- df %>% dplyr::filter(is.na(transcript_id))

df <- df %>% dplyr::filter(!is.na(transcript_id))

```

Note: in the report we refer to a uniquely mapped peptide as a 'UMP'. These are peptides which only mapped to one protein/ORF.

### Summary of peptides, transcripts and ORFs

```{r, eval=TRUE, echo=FALSE}

peptotal <- length(unique(df$peptide))

uniqpeptides <- nrow(df %>% group_by(peptide) %>%
            dplyr::filter(length(unique(PID))==1) %>%
            slice_head(n=1))

txtotal <- length(unique(df$transcript_id))

txuniqtotal <- nrow(df %>% group_by(transcript_id) %>%
            slice_head(n=1) %>% ungroup() %>% 
            dplyr::filter(transcript_identified == TRUE))

genetotal <- length(unique(df$gene_id))

geneuniqtotal <- nrow(df %>% group_by(peptide) %>%
            dplyr::filter(length(unique(gene_id)) == 1) %>%
            slice_head(n=1) %>% ungroup() %>% 
            group_by(gene_id) %>% slice_head(n=1))

orftotal <- length(unique(df$PID))

orfuniqtotal <- nrow(df %>% group_by(PID) %>%
            slice_head(n=1) %>% ungroup() %>% 
            dplyr::filter(orf_identified == TRUE))

```

Total peptides: `r paste0(peptotal)`

Uniquely mapped peptides (UMPs): `r paste0(uniqpeptides)`

Total transcripts: `r paste0(txtotal)`

Total transcripts with ≥1 UMP: `r paste0(txuniqtotal)`

Total genes: `r paste0(genetotal)`

Total genes with ≥1 UMP: `r paste0(geneuniqtotal)`

Total ORFs: `r paste0(orftotal)`

Total ORFs with ≥1 UMP: `r paste0(orfuniqtotal)`



### Summary of peptides mapped to contaminants, decoys or that weren't mapped to genomic loci

```{r, eval=TRUE, echo=FALSE}


bad_rev <- bad_pep %>% dplyr::filter(startsWith(PID, "rev"))
if (nrow(bad_rev) == 0) {
  bad_rev_num <- 0
} else {
  bad_rev_num <- length(unique(bad_rev$peptide))
}

bad_contam <- bad_pep %>% dplyr::filter(startsWith(PID, "sp"))
if (nrow(bad_contam) == 0) {
  bad_contam_num <- 0
} else {
  bad_contam_num <- length(unique(bad_contam$peptide))
}

bad_multi_loc <- bad_pep %>% dplyr::filter(!startsWith(PID, "sp") & !startsWith(PID, "rev"))
if (nrow(bad_multi_loc) == 0) {
  bad_multi_loc_num <- 0
} else {
  bad_multi_loc_num <- length(unique(bad_multi_loc$peptide))
}

```

Total peptides mapped to decoys: `r paste0(bad_rev_num)`
Total peptides mapped to contaminants: `r paste0(bad_contam_num)`
Total other peptides that weren't mapped (likely belong to ORFs with multiple genomic locations): `r paste0(bad_multi_loc_num)`



### Peptide-level data

```{r, eval=TRUE, echo=FALSE}

b <- nrow(df %>% group_by(peptide) %>%
            dplyr::filter(length(unique(PID))==1 & length(unique(transcript_id))==1) %>%
            slice_head(n=1))
c <- nrow(df %>% group_by(peptide) %>%
            dplyr::filter(length(unique(PID))==1) %>%
            slice_head(n=1))
d <- nrow(df %>% group_by(peptide) %>%
            dplyr::filter(length(unique(gene_id))==1) %>%
            slice_head(n=1))
e <- nrow(df %>% group_by(peptide) %>%
            dplyr::filter(length(unique(PID))>1 & length(unique(gene_id))>1) %>%
            slice_head(n=1))
f <- nrow(df %>% group_by(peptide) %>%
          dplyr::filter(number_exons > 1) %>%
          slice_head(n=1))

# Make stats
item <- c(
  "peptides mapped to a single transcript",
  "peptides mapped to a single ORF (UMPs)",
  "peptides mapped to a single gene",
  "peptides mapped to multiple ORFs from multiple genes",
  "peptides spanning splice junctions")

total <- c(b,c,d,e,f)
stats <- data.frame(item, total)

knitr::kable(stats, format = "html", col.names=c(" ", ""))

# Export overall stats file
#write.table(stats, file = paste0(output,"_stats.csv"), sep=",", quote=F, col.names=F, row.names=F)

```



### Transcript-level data (with ≥1 UMP)

```{r, eval=TRUE, echo=FALSE}

k <- nrow(df %>% group_by(transcript_id) %>%
            slice_head(n=1) %>% ungroup() %>% 
            dplyr::filter(transcript_identified == TRUE & !startsWith(transcript_id, "Bambu")))
l <- nrow(df %>% group_by(transcript_id) %>%
            slice_head(n=1) %>% ungroup() %>% 
            dplyr::filter(transcript_identified == TRUE & startsWith(transcript_id, "Bambu")))
m <- nrow(df %>% group_by(transcript_id) %>%
            slice_head(n=1) %>%
            ungroup() %>% dplyr::filter(transcript_identified == TRUE & startsWith(transcript_id, "Bambu")) %>%
            group_by(PID) %>% dplyr::filter(!startsWith(PID, "ORF")) %>% slice_head(n=1))
n <- nrow(df %>% group_by(transcript_id) %>%
            slice_head(n=1) %>%
            ungroup() %>% dplyr::filter(transcript_identified == TRUE & startsWith(transcript_id, "Bambu")) %>%
            group_by(PID) %>% dplyr::filter(startsWith(PID, "ORF")) %>% slice_head(n=1))

# Make stats
item <- c(
  "known isoforms",
  "novel isoforms",
  "novel isoforms encoding known ORFs",
  "novel isoforms encoding novel ORFs")

total <- c(k,l,m,n)
stats <- data.frame(item, total)

knitr::kable(stats, format = "html", col.names=c(" ", ""))

# Export overall stats file
#write.table(stats, file = paste0(output,"_stats.csv"), sep=",", quote=F, col.names=F, row.names=F)

```



### Gene-level data (with ≥1 UMP)

```{r, eval=TRUE, echo=FALSE}

h1 <- nrow(df %>% group_by(gene_id) %>%
             dplyr::filter(peptide_ids_gene == TRUE) %>% 
            slice_head(n=1) %>%
            ungroup() %>% dplyr::filter(!startsWith(gene_id, "Bambu")))
i1 <- nrow(df %>% group_by(gene_id) %>%
             dplyr::filter(peptide_ids_gene == TRUE) %>% 
            slice_head(n=1) %>%
            ungroup() %>% dplyr::filter(startsWith(gene_id, "Bambu")))

# Make stats
item <- c(
  "known genes identified by ≥1 UMP",
  "novel genes identified by ≥1 UMP")

total <- c(h1,i1)
stats <- data.frame(item, total)

knitr::kable(stats, format = "html", col.names=c(" ", ""))

# Export overall stats file
#write.table(stats, file = paste0(output,"_stats.csv"), sep=",", quote=F, col.names=F, row.names=F)


```



### ORF-level data (with ≥1 UMP)

```{r, eval=TRUE, echo=FALSE}

h <- nrow(df %>% group_by(PID) %>%
            slice_head(n=1) %>%
            ungroup() %>% dplyr::filter(orf_identified == TRUE & !startsWith(PID, "ORF")))
i <- nrow(df %>% group_by(PID) %>%
            slice_head(n=1) %>%
            ungroup() %>% dplyr::filter(orf_identified == TRUE & startsWith(PID, "ORF")))

# Make stats
item <- c(
  "known ORFs identified by ≥1 UMP",
  "novel ORFs identified by ≥1 UMP")

total <- c(h,i)
stats <- data.frame(item, total)

knitr::kable(stats, format = "html", col.names=c(" ", ""))

# Export overall stats file
#write.table(stats, file = paste0(output,"_stats.csv"), sep=",", quote=F, col.names=F, row.names=F)


```



### Figures

#### Peptides by mapping status

```{r, eval=TRUE, echo=FALSE}

plot1 <- data.frame(
  metric = c("peptides", "peptide ids gene", "UMP"),
  total = c(peptotal, d, c)
  )

plot1$metric <- factor(plot1$metric,
                       levels=c("peptides", "peptide ids gene", "UMP"))

pepplot <- ggplot(plot1) +
  geom_bar(aes(x=metric, y=total, fill=metric), stat="identity") +
  scale_fill_manual(
    values = c("peptides" = "#D3D3D3", "peptide ids gene" = "#818589", "UMP" = "orangered2")
    ) +
  ylab("Total") +
  xlab("Category") +
  theme_bw() +
  theme(legend.position = "none")

print(pepplot)

```



#### UMPs by number of exons spanned

```{r, eval=TRUE, echo=FALSE}

summary_df <- df %>% 
  group_by(peptide) %>% 
  dplyr::filter(length(unique(PID)) == 1) %>%
  ungroup() %>% 
  group_by(number_exons) %>%
  summarise(total = n())

exons_plot <- ggplot(summary_df) +
  geom_bar(aes(x=number_exons, y=total, fill=number_exons), stat="identity") +
  ylab("Total") +
  xlab("Number of exons peptide spans") +
  theme_bw() +
  theme(legend.position = "none")

print(exons_plot)

```



#### Novel ORFs (with ≥1 UMP) by localisation

```{r, eval=TRUE, echo=FALSE}

summary_df <- df %>% group_by(peptide) %>% 
            dplyr::filter(length(unique(PID)) == 1) %>%
            ungroup() %>% 
            group_by(PID) %>% slice_head(n=1) %>% ungroup() %>% 
            dplyr::filter(startsWith(PID, "ORF")) %>%
            group_by(localisation) %>%
            summarise(total = n())

plot_orf_local <- ggplot(summary_df) +
  geom_bar(aes(x=localisation, y=total, fill=localisation), stat="identity") +
  ylab("Total") +
  xlab("ORF localisation") +
  theme_bw() +
  theme(legend.position = "none")

print(plot_orf_local)

```



#### Novel ORFs (with ≥1 UMP) by transcript biotype

```{r, eval=TRUE, echo=FALSE}

summary_df <- df %>% group_by(peptide) %>% 
            dplyr::filter(length(unique(PID)) == 1) %>%
            ungroup() %>% 
            group_by(PID) %>% slice_head(n=1) %>% ungroup() %>% 
            dplyr::filter(startsWith(PID, "ORF")) %>%
            group_by(transcript_biotype) %>%
            summarise(total = n())

plot_orf_biotype <- ggplot(summary_df) +
  geom_bar(aes(x=transcript_biotype, y=total, fill=transcript_biotype), stat="identity") +
  ylab("Total") +
  xlab("Transcript biotype") +
  theme_bw() +
  theme(legend.position = "none", 
        axis.text.x = element_text(angle = -45, hjust = 0, vjust = 1))

print(plot_orf_biotype)

```



#### Lengths of detected ORFs (known and novel)

```{r, eval=TRUE, echo=FALSE}

summary_df <- df %>% group_by(peptide) %>% 
            dplyr::filter(length(unique(PID)) == 1) %>%
            ungroup() %>% 
            group_by(PID) %>% slice_head(n=1) %>% ungroup()

# plot_orf_lengths <- ggplot(summary_df) +
#   geom_boxplot(aes(x=orf_type, y=protein_length, fill=orf_type), outlier.shape=NA) +
#   ylab("Length") +
#   xlab("ORF type") +
#   theme_bw() +
#   ylim(0,1000) +
#   theme(legend.position = "none", 
#         axis.text.x = element_text(angle = -45, hjust = 0, vjust = 1))

plot_orf_lengths <- ggplot(summary_df) +
  geom_density(aes(x=protein_length, y=..scaled.., color=orf_type)) +
  ylab("Scaled density") +
  xlab("ORF Length (AA)") +
  theme_bw() +
  xlim(20,1000)

print(plot_orf_lengths)

```



