---
title: "GenomeProt Report"
date: "`r format(Sys.Date())`"
output: html_document
layout: single
params:
  directory:
    value: x
  file:
    value: x
---


```{r setup, include=FALSE, warning=FALSE}
library(ggplot2)
library(gplots)
library(dplyr)
library(RColorBrewer)
library(data.table)
library(scales)
library(openxlsx)

options(scipen=999)

df <- read_tsv(paste0(params$directory, "/", params$file))

#df <- read_tsv("/home/ubuntu/data_transfer/peptide_info.tsv")
wb<-createWorkbook()
addWorksheet(wb,"ORF_by_location")
addWorksheet(wb,"ORF_by_tx_biotype")


bad_pep <- df %>% dplyr::filter(is.na(transcript_id))

df <- df %>% dplyr::filter(!is.na(transcript_id))

```

##### Definitions:
ORF: Open reading frame.

Canonical ORF: ORF annotated in UniProt or RefSeq databases.

Noncanonical ORF (ncORF): ORF not annotated in UniProt or RefSeq databases.

UMP (Uniquely Mapped Peptide): Peptide uniquely mapped to a single ORF or protein.

Peptide evidence: Feature supported by any mapped peptide.

Transcript-specific or gene-specific peptide: Peptide uniquely mapped to a specific transcript or gene.

Peptide-confirmed: Feature distinguished by transcript-specific or gene-specific peptides.

### Summary of peptides, transcripts and ORFs

```{r, eval=TRUE, echo=FALSE}

peptotal <- length(unique(df$peptide))

uniqpeptides <- nrow(df %>% group_by(peptide) %>%
            dplyr::filter(length(unique(PID))==1) %>%
            slice_head(n=1))

txtotal <- length(unique(df$transcript_id))

txuniqtotal <-nrow(df%>%dplyr::filter(transcript_identified == TRUE)%>%dplyr::select("transcript_id")%>%unique())
  
  
  nrow(df%>%dplyr::filter(peptide_ids_transcript == TRUE)%>%dplyr::select("peptide")%>%unique())
  

genetotal <- length(unique(df$gene_id))

geneuniqtotal <- df %>%dplyr::filter(gene_identified==TRUE)%>%dplyr::select("gene_id")%>%unique()%>%NROW

orftotal <- unique(df$accession)%>%NROW

orfuniqtotal <- df %>% dplyr::filter(orf_identified == TRUE)%>%dplyr::select(accession)%>%unique()%>%NROW


```

Total Peptides: `r paste0(peptotal)`

UMPs: `r paste0(uniqpeptides)`

Transcripts with peptide evidence: `r paste0(txtotal)`

Peptide-confirmed transcripts: `r paste0(txuniqtotal)`

Genes with peptide evidence: `r paste0(genetotal)`

Peptide-confirmed genes: `r paste0(geneuniqtotal)`

Proteins/ORFs with peptide evidence: `r paste0(orftotal)`

Proteins/ORFs with UMPs: `r paste0(orfuniqtotal)`



### Summary of peptides mapped to contaminants, decoys or that weren't mapped to genomic loci

```{r, eval=TRUE, echo=FALSE}

bad_rev <- bad_pep %>% dplyr::filter(startsWith(PID, "rev"))
if (nrow(bad_rev) == 0) {
  bad_rev_num <- 0
} else {
  bad_rev_num <- length(unique(bad_rev$peptide))
}

bad_contam <- bad_pep %>% dplyr::filter(startsWith(PID, "sp"))
if (nrow(bad_contam) == 0) {
  bad_contam_num <- 0
} else {
  bad_contam_num <- length(unique(bad_contam$peptide))
}

bad_multi_loc <- bad_pep %>% dplyr::filter(!startsWith(PID, "sp") & !startsWith(PID, "rev"))
if (nrow(bad_multi_loc) == 0) {
  bad_multi_loc_num <- 0
} else {
  bad_multi_loc_num <- length(unique(bad_multi_loc$peptide))
}

```

Total peptides mapped to decoys: `r paste0(bad_rev_num)`

Total peptides mapped to contaminants: `r paste0(bad_contam_num)`

Total other peptides that weren't mapped (likely belong to ORFs with multiple genomic locations): `r paste0(bad_multi_loc_num)`



### Peptide-level data

```{r, eval=TRUE, echo=FALSE}

#total uniquely mapped peptides
a <- df%>%dplyr::filter(peptide_ids_orf==TRUE)%>%dplyr::select(peptide)%>%unique()%>%NROW
#Transcript-specific peptides   
b <- df %>% dplyr::filter(peptide_ids_transcript==TRUE)%>%dplyr::select(peptide)%>%unique()%>%NROW
#Gene-specific peptides            
c <- df %>% dplyr::filter(peptide_ids_gene==TRUE)%>%dplyr::select(peptide)%>%unique()%>%NROW
#Peptides spanning splice junctions 
d <- df%>%dplyr::filter(number_of_mapped_exons > 1)%>%dplyr::select(peptide)%>%unique()%>%NROW
  
#Peptides mapped to multiple ORFs from multiple genes 	
e <- nrow(df %>% group_by(peptide) %>%
            dplyr::filter(length(unique(PID))>1 & length(unique(gene_id))>1) %>%
            slice_head(n=1))

# Make stats
item <- c(
  "Uniquely mapped peptides ",
  "Transcript-specific peptides ",
  "Gene-specific peptides ",
  "Peptides spanning splice junctions ",
  "Peptides mapped to multiple ORFs from multiple genes "
  )

total <- c(a,b,c,d,e)
stats <- data.frame(item, total)

knitr::kable(stats, format = "html", col.names=c(" ", ""))

# Export overall stats file
#write.table(stats, file = paste0(output,"_stats.csv"), sep=",", quote=F, col.names=F, row.names=F)

```



### Transcript-level data 

```{r, eval=TRUE, echo=FALSE}

#Known transcripts with peptide evidence 
f <- df %>% filter(!startsWith(transcript_id, "Bambu"))%>%dplyr::select(transcript_id)%>%unique()%>%NROW
#Novel transcripts with peptide evidence 
g <- df %>% filter(startsWith(transcript_id, "Bambu"))%>%dplyr::select(transcript_id)%>%unique()%>%NROW
  
#Known transcripts encoding ORFs with UMPs 
h <- df %>% dplyr::filter(orf_identified == TRUE)%>%filter(!startsWith(transcript_id, "Bambu"))%>%dplyr::select(transcript_id)%>%unique()%>%NROW

#Novel transcripts encoding ORFs with UMPs 
i <- df %>% dplyr::filter(orf_identified == TRUE)%>%filter(startsWith(transcript_id, "Bambu"))%>%dplyr::select(transcript_id)%>%unique()%>%NROW

#Peptide-confirmed known transcripts 
j <- df %>% dplyr::filter(transcript_identified == TRUE & !startsWith(transcript_id, "Bambu"))%>%dplyr::select(transcript_id)%>%unique()%>%NROW
  
#Peptide-confirmed novel transcripts
k <-df %>% dplyr::filter(transcript_identified == TRUE & startsWith(transcript_id, "Bambu"))%>%dplyr::select(transcript_id)%>%unique()%>%NROW
  
#Peptide-confirmed novel transcripts encoding annotated ORFs with UMPs
l <- df %>% dplyr::filter(transcript_identified == TRUE & startsWith(transcript_id, "Bambu"))%>%dplyr::filter(!startsWith(accession, "ORF_"))%>%dplyr::select(accession)%>%unique()%>%NROW

#Peptide-confirmed novel transcripts encoding unannotated ORFs with UMPs 	
m <- df %>% dplyr::filter(transcript_identified == TRUE & startsWith(transcript_id, "Bambu"))%>%dplyr::filter(startsWith(accession, "ORF_"))%>%dplyr::select(accession)%>%unique()%>%NROW


# Make stats
item <- c(
  "Annotated (known) transcripts with peptide evidence ",
  "Unannotated (novel) transcripts with peptide evidence ",
  "Annotated transcripts encoding ORFs with UMPs ",
  "Unannotated transcripts encoding ORFs with UMPs ",
  "Peptide-confirmed annotated transcripts ",
  "Peptide-confirmed unannotated transcripts ",
  "Peptide-confirmed unannotated transcripts encoding canonical ORFs with UMPs ",
  "Peptide-confirmed unannotated transcripts encoding noncanonical ORFs with UMPs ")

total <- c(f,g,h,i,j,k,l,m)
stats <- data.frame(item, total)

knitr::kable(stats, format = "html", col.names=c(" ", ""))

# Export overall stats file
#write.table(stats, file = paste0(output,"_stats.csv"), sep=",", quote=F, col.names=F, row.names=F)

```



### Protein/ORF-level data

```{r, eval=TRUE, echo=FALSE}
#Annotated ORFs with peptide evidence
n <-df %>% dplyr::filter(!startsWith(accession, "ORF_"))%>%dplyr::select(accession)%>%unique()%>%NROW

#Unannotated ORFs with peptide evidence 	 
o <-df %>% dplyr::filter(startsWith(accession, "ORF_"))%>%dplyr::select(accession)%>%unique()%>%NROW

#Annotated ORFs with UMPs
p <-df %>% dplyr::filter(orf_identified == TRUE & !startsWith(accession, "ORF_"))%>%dplyr::select(accession)%>%unique()%>%NROW
  
#Unannotated ORFs with UMPs 
q <-df %>% dplyr::filter(orf_identified == TRUE & startsWith(accession, "ORF_"))%>%dplyr::select(accession)%>%unique()%>%NROW
  


# Make stats
item <- c(
  "Canonical ORFs with peptide evidence ",
  "Noncanonical ORFs with peptide evidence ",
  "Canonical ORFs with UMPs ",
  "Noncanonical ORF with UMPs ")

total <- c(n,o,p,q)
stats <- data.frame(item, total)

knitr::kable(stats, format = "html", col.names=c(" ", ""))

# Export overall stats file
#write.table(stats, file = paste0(output,"_stats.csv"), sep=",", quote=F, col.names=F, row.names=F)

```



### Figures

#### Peptides by mapping status (Total peptides: `r peptotal`)

```{r, eval=TRUE, echo=FALSE, message = FALSE, warning = FALSE}

plot1 <- data.frame(
  Category = c("Shared peptides", "Uniquely mapped peptides"),
  count = c(peptotal-a, a)
  )


plot1$percent <- round(100 * plot1$count / sum(plot1$count), 1)

plot1$label<-paste0(plot1$count, "(",plot1$percent, "%)")

pepplot<-ggplot(plot1, aes(x = "", y = count, fill = Category)) +
  geom_col(width = 1,color= "white") +
  coord_polar(theta = "y") +
  geom_text(aes(label = label),
            position = position_stack(vjust = 0.5),size=5) +
  theme_void()+labs(fill="Category")+
  scale_fill_manual(values = c("Shared peptides"="#D3D3D3", "Uniquely mapped peptides" ="lightgreen"))


suppressWarnings(print(pepplot))


```



#### Number of UMPs per ORF (shown for >10 ORFs)

```{r, eval=TRUE, echo=FALSE, message = FALSE, warning = FALSE}

umps_per_protein <- df %>% 
  dplyr::filter(orf_identified == TRUE) %>% 
  group_by(accession) %>%
  summarise(number_ump = length(unique(peptide))) %>% 
  ungroup() %>% 
  mutate(orf_type = case_when(
    startsWith(accession, "ORF") ~ "unannotated",
    TRUE ~ "annotated"
  ))

tallied <- umps_per_protein %>% 
  group_by(orf_type, number_ump) %>% 
  summarise(n = n()) %>% 
  dplyr::filter(n>10)%>%rename("ORF type"=orf_type)

ump_per_protein_plot <- ggplot(tallied) +
  geom_bar(aes(x=number_ump, y=n, fill=`ORF type`), position="stack", stat="identity") +
  ylab("Number of ORFs") +
  xlab("Number of UMPs per ORF") +
  scale_x_continuous(breaks = pretty_breaks()) +
  scale_y_continuous(breaks = pretty_breaks()) +
  scale_fill_manual(values=c("unannotated" = "deepskyblue3", "annotated" = "darkgreen")) +
  theme_bw()

suppressWarnings(print(ump_per_protein_plot))

```



#### Peptides by number of exons spanned

```{r, eval=TRUE, echo=FALSE, message = FALSE, warning = FALSE}

summary_df <- df %>% 
  group_by(peptide) %>% 
  slice_head(n=1) %>% 
  ungroup() %>% 
  group_by(number_of_mapped_exons) %>%
  summarise(total = n())

exons_plot <- ggplot(summary_df) +
  geom_bar(aes(x=number_of_mapped_exons, y=total, fill=number_of_mapped_exons), stat="identity") +
  ylab("Number of peptides") +
  xlab("Number of exons spanned") +
  theme_bw() +
  theme(legend.position = "none")

suppressWarnings(print(exons_plot))

```



#### Noncanonical ORFs with UMPs by localisation

```{r, eval=TRUE, echo=FALSE, message = FALSE, warning = FALSE}

summary_df <- df %>%dplyr::filter(peptide_ids_orf==TRUE & startsWith(accession, "ORF"))%>%dplyr::select(accession,localisation)%>%unique()%>%
  group_by(localisation) %>%summarise(Count = n())

writeData(wb,sheet = "ORF_by_location",x=summary_df)

plot_orf_local <- ggplot(summary_df) +
  geom_bar(aes(x=localisation, y=Count, fill=localisation), stat="identity") +
  ylab("Number of ncORFs") +
  xlab("Localisation") +
  scale_fill_brewer(palette = "Dark2") +
  theme_bw() +
  theme(legend.position = "none")

suppressWarnings(print(plot_orf_local))

```



#### Noncanonical ORFs with UMPs by simplified transcript biotype

```{r, eval=TRUE, echo=FALSE, message = FALSE, warning = FALSE}

summary_df <- df%>%dplyr::filter(orf_identified == TRUE & startsWith(accession, "ORF"))%>%select(accession,simplified_biotype)%>%unique()%>%
  group_by(simplified_biotype) %>%summarise(Count = n())

writeData(wb,sheet = "ORF_by_tx_biotype",x=summary_df)

summary_df <- df%>%dplyr::filter(orf_identified == TRUE & startsWith(accession, "ORF"))%>%dplyr::select(accession,simplified_biotype)%>%unique()%>%
  group_by(simplified_biotype) %>%summarise(Count = n())

writeData(wb,sheet = "ORF_by_tx_biotype",x=summary_df)


plot_orf_biotype <- ggplot(summary_df) +
  geom_bar(aes(x=simplified_biotype, y=Count, fill=simplified_biotype), stat="identity") +
  ylab("Number of ncORFs") +
  xlab("Transcript biotype") +
  scale_fill_brewer(palette = "Set3") +
  theme_bw() +
  theme(legend.position = "none", 
        axis.text.x = element_text(angle = -45, hjust = 0, vjust = 1))

suppressWarnings(print(plot_orf_biotype))

```



#### Lengths of annotated and unannotated ORFs (with UMPs)

```{r, eval=TRUE, echo=FALSE, message = FALSE, warning = FALSE}

summary_df <- df %>% dplyr::filter(orf_identified == TRUE) %>% 
            dplyr::select(accession,protein_length,orf_type) %>%unique()%>%
  rename("ORF type"=orf_type)

plot_orf_lengths <- ggplot(summary_df) +
  geom_density(aes(x=protein_length, y=..scaled.., color=`ORF type`)) +
  ylab("Scaled density") +
  xlab(paste0("ORF Length (", min_orf_length, "-1000 AA)")) +
  scale_color_manual(values = c("annotated" = "darkgreen", "unannotated" = "deepskyblue3")) +
  theme_bw() +
  xlim(min_orf_length,1000)

suppressWarnings(print(plot_orf_lengths))

```



```{r, eval=TRUE, echo=FALSE, message = FALSE, warning = FALSE}

#save stats
saveWorkbook(wb,paste0(params$directory,"/ncORF_stats.xlsx"),overwrite = TRUE)
# save all PDFs
pdf(paste0(params$directory,"/report_images/peptides-by-mapping-status.pdf"), w=6, h=4)
suppressWarnings(print(pepplot))
invisible(dev.off())

pdf(paste0(params$directory,"/report_images/umps-per-orf.pdf"), w=6, h=4)
suppressWarnings(print(ump_per_protein_plot))
invisible(dev.off())

pdf(paste0(params$directory,"/report_images/exons-spanned-by-peptides.pdf"), w=6, h=4)
suppressWarnings(print(exons_plot))
invisible(dev.off())

pdf(paste0(params$directory,"/report_images/unannotated-orf-localisations.pdf"), w=6, h=4)
suppressWarnings(print(plot_orf_local))
invisible(dev.off())

pdf(paste0(params$directory,"/report_images/unannotated-orf-tx-biotypes.pdf"), w=6, h=4)
suppressWarnings(print(plot_orf_biotype))
invisible(dev.off())

pdf(paste0(params$directory,"/report_images/lengths-annotated-v-unannotated-orfs.pdf"), w=6, h=4)
suppressWarnings(print(plot_orf_lengths))
invisible(dev.off())

```





